### ISSUE TITLE
Migration not considering all objects when objectClasses are set

### ISSUE THREAD

**Issue by SystemKeeper** [realm/realm-cocoa#7030]
When accessing realm from an app extension, we narrow down the available objects (because of memory limits by apple) doing something like this
``` 
RLMRealmConfiguration *configuration = [RLMRealmConfiguration defaultConfiguration];
configuration.fileURL = databaseURL;
configuration.schemaVersion = schemaVersion;
configuration.objectClasses = @[RequiredClass.class];       
configuration.migrationBlock = ^(RLMMigration *migration, uint64_t oldSchemaVersion) {
};
[RLMRealmConfiguration setDefaultConfiguration:configuration];
```
This works fine, unless we increase the schemaVersion and therefore trigger an migration. Then the schemaVersion gets updated correctly, but classes which are not part of `objectClasses` are not migrated, resulting in a crash, when opening the main app.

## Expected Results
Not absolutely sure what to expect here. On one hand it makes sense to only process objects which are explicitly set in `objectClasses`, but on the other hand I'd expect the whole schema to be migrated (which might be problematic again, memory-wise?!).
At least a hint about this in the migration section (https://realm.io/docs/objc/latest#migrations) would be very useful, because I don't think that missing migration can be triggered again without increasing the schemaVersion again.

## Actual Results
The objects which are not specified in `objectClasses` are not migrated, but the schemaVersion is increased.

As `objectClasses` internally set a custom schema, this makes sense: https://github.com/realm/realm-cocoa/blob/fc50ee023e79b00931df089dae5e45b9d94e87ac/Realm/RLMRealm.mm#L465

and then

https://github.com/realm/realm-cocoa/blob/fc50ee023e79b00931df089dae5e45b9d94e87ac/Realm/RLMRealm.mm#L487-L490

## Steps for others to Reproduce

1. Have `Class1` and `Class2`
2. Increase `schemaVersion` and add a property to `Class1`, but set `objectClasses` only to `Class2`
3. Run the app -> check that the `schemaVersion` actually increased.
4. Add `Class1` to `objectClasses` 
5. Run the app -> exception because of a missing migration.

## Version of Realm and Tooling
```
ProductName:	Mac OS X (iOS App on iOS 14.2)
ProductVersion:	10.15.5
BuildVersion:	19F101

/Applications/Xcode.app/Contents/Developer
Xcode 12.2
Build version 12B45b

/usr/local/bin/pod
1.10.0
Realm (10.1.4)
Realm (= 10.1.4)

/bin/bash
GNU bash, version 3.2.57(1)-release (x86_64-apple-darwin19)

/usr/local/bin/carthage
0.36.0
(not in use here)

/usr/bin/git
git version 2.24.3 (Apple Git-128)
```

**Comment by tgoyne** [comment]
This is working as intended, but does indeed seem like a problem.

The primary memory usage benefit of specifying objectClasses comes from that it lets us skip calling objc_copyClassList() (which forces a bunch of lazily-loaded things to be immediately loaded), and actually limiting the set of classes used is relatively minor in comparison. If you haven't, it'd be worth measuring the memory usage of including all of your types in objectClasses and seeing if it's an issue.

If you really do need to use just a minimal subset then it seems like our schema version concept can't work well for this. An awkward workaround would be to set the schema version one higher in the app than in the extension. When opening the Realm in the extension, you'd need to have error handling which checks if the error was that the file has the app's schema version, and if so open the Realm with that version instead. You might be tempted to use schemaVersionAtURL: instead of handling the error, but that'd be a race condition if the extension and app launched at the same time and wouldn't be faster.

**Comment by SystemKeeper** [comment]
We indeed saw some crash reports in app extensions. That stopped after using objectClasses, but I do not have any numbers at hand. After reading your explanation I'm wondering if there was a different root cause to the crashes we saw. How memory intensive is a migration? Could it be that the app extension would crash if it attempts to do a migration of all objects?

Using two different schema versions sound indeed like a awkward workaround. I guess that will get us in trouble at some point... Another way I could think of is using schemaVersionAtURL: in the app extension and just cancel the request there (I guess that shouldn't be a problem in case of race conditions then?!). A user then needs to open the main app at least one time after an update, but we would have consistent migrations.

**Comment by tgoyne** [comment]
The memory-intensive thing is the automatic discovery of RLMObject subclasses, not the migration. Setting objectClasses to the complete list of types used by the app greatly reduces the peak memory usage on startup, while setting it to a smaller subset of the types is a relatively small decrease in memory usage.

**Comment by SystemKeeper** [comment]
So if I understood you correctly, when we specify all our RLMObjects like this
```
configuration.objectClasses = @[Class1.class, Class2.class, Class3.class, ...];
```
it's less memory intensive then omitting the `objectClasses` alltogether and letting realm discover all classes? That would explain why we did see memory crashes in app extensions, because we didn't specify `objectClasses` when we wanted to use all of them.

**Comment by tgoyne** [comment]
Yes, listing all of them reduces peak memory compared to not setting objectClasses at all.
